parsl/serialize/facade.py

```diff
@@ -0,0 +1,174 @@
+
+    def __new__(cls, *args, **kwargs):
+        """ Make ParslSerializer a singleton using the fact that there's only one instance
+        of a class variable
+        """
+        if ParslSerializer.__singleton_instance is None:
+            ParslSerializer.__singleton_instance = object.__new__(cls, *args, **kwargs)
+        return ParslSerializer.__singleton_instance
+
+    def __init__(self):
+        """ Instantiate the appropriate classes
+        """
+        headers = list(METHODS_MAP_CODE.keys()) + list(METHODS_MAP_DATA.keys())
+        self.header_size = len(headers[0])
+
+        self.methods_for_code = {}
+        self.methods_for_data = {}
+
+        for key in METHODS_MAP_CODE:
+            self.methods_for_code[key] = METHODS_MAP_CODE[key]()
+            self.methods_for_code[key].enable_caching(maxsize=128)
+
+        for key in METHODS_MAP_DATA:
+            self.methods_for_data[key] = METHODS_MAP_DATA[key]()
+
+    def _list_methods(self):
+        return self.methods_for_code, self.methods_for_data
+
+    def pack_apply_message(self, func, args, kwargs, buffer_threshold=1e6):
+        """Serialize and pack function and parameters
+
+        Parameters
+        ----------
+
+        func: Function
+            A function to ship
+
+        args: Tuple/list of objects
+            positional parameters as a list
+
+        kwargs: Dict
+            Dict containing named parameters
+
+        buffer_threshold: Ignored
+            Limits buffer to specified size in bytes. Default is 1e6 bytes.
+        """
+        b_func = self.serialize(func, buffer_threshold=buffer_threshold)
+        b_args = self.serialize(args, buffer_threshold=buffer_threshold)
+        b_kwargs = self.serialize(kwargs, buffer_threshold=buffer_threshold)
+        packed_buffer = self.pack_buffers([b_func, b_args, b_kwargs])
+        return packed_buffer
+
+    def unpack_apply_message(self, packed_buffer, user_ns=None, copy=False):
+        """ Unpack and deserialize function and parameters
+
+        """
+        return [self.deserialize(buf) for buf in self.unpack_buffers(packed_buffer)]
+
+    def serialize(self, obj, buffer_threshold=1e6):
+        """ Try available serialization methods one at a time
+
+        Individual serialization methods might raise a TypeError (eg. if objects are non serializable)
+        This method will raise the exception from the last method that was tried, if all methods fail.
+        """
+        serialized = None
+        serialized_flag = False
+        last_exception = None
+        if callable(obj):
+            for method in self.methods_for_code.values():
+                try:
+                    serialized = method.serialize(obj)
+                    # We attempt a deserialization to make sure both work.
+                    method.deserialize(serialized)
+                except Exception as e:
+                    logger.exception(f"Serialization method: {method} did not work")
```

Comment:
I think this is going to log an `ERROR` level log for every fallover-to-next-serializer serialisation that happens (multiple times per app invocation perhaps), even though no error condition exists. It should be debug or info level.